docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,309 @@
+ManagedCluster after creation (if allowed by AKS) because CAPZ will not override user-provided fields for
+which it does not have its own opinion on ASO resources.
+
+#### Decision
+
+TBD
+
+### Security Model
+
+Overall, none of the approaches outlined above change what data is ultimately represented in the API, only the
+higher-level shape of the API. That means there is no further transport or handling of secrets or other
+sensitive information beyond what CAPZ already does.
+
+### Risks and Mitigations
+
+Increasing CAPZ's reliance on ASO and exposing ASO to users at the API level further increases the risk
+[previously discussed](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/ce3c130266b23a8b67aa5ef9a21f257ff9e6d63e/docs/proposals/20230123-azure-service-operator.md?plain=1#L169)
+that since ASO has not yet been proven to be as much of a staple as the other projects that manage
+infrastructure on Azure, ASO's lifespan may be more limited than others. If ASO were to sunset while CAPZ
+still relies on it, CAPZ would have to rework its APIs. This risk is mitigated by the fact that no
+announcements have yet been made regarding ASO's end-of-life and the project continues to be very active.
+
+## Alternatives
+
+The main alternative to leveraging ASO to expose the entire AKS API from CAPZ would be for CAPZ to build its
+own code generation pipeline which takes as input the Azure API specs and ultimately produces definitions for
+AzureManagedControlPlane and AzureManagedMachinePool and mappings to ASO's ManagedCluster and
+ManagedClustersAgentPool. Such a solution would likely end up being too similar to what ASO already provides
+for the effort to be worthwhile.
+
+## Upgrade Strategy
+
+Each of the [options above](#api-design-options) suggest additive, backwards-compatible API changes. Users
+will not need to perform any action after upgrading to the first version of CAPZ that implements this
+proposal.
+
+The new API field allowing the entire ASO resource to be defined in the CAPZ API will overlap with several
+existing CAPZ API fields. e.g. with [Option 4], AzureManagedControlPlane's `spec.location` maps to the same
+ASO API field as `spec.managedClusterSpec.location`. If both are defined, then the `spec.managedCluster.*`
+field takes precedence. If only `spec.location` is defined, its value will still be used to construct the
+desired state for the ManagedCluster.
+
+## Additional Details
+
+### Test Plan
+
+Existing end-to-end tests will verify that CAPZ's current behavior does not regress. New tests will verify
+that the new API fields proposed here behave as expected.
+
+### Graduation Criteria
+
+In CAPZ version `1.N` which first implements this proposal, setting any new CAPZ API fields introduced by this
+proposal will not be allowed without enabling an `AKSASOAPI` feature flag in CAPZ. The feature flag will be
+required throughout `1.N+2` and dropped for `1.N+3` where the feature will be enabled by default.
```

Comment:
One thing I don't think I mentioned anywhere in the proposal is what to do with the fields that already exist that will be made redundant. Should we deprecate them? Can we eventually remove them, or is that too disruptive without changing the CRD API version?

Reply:
Leaving this here for reference: https://kubernetes.io/docs/reference/using-api/deprecation-policy/